File: vol1/ch11/LeetCode370.cpp
class Solution {
public:
    vector<int> getModifiedArray(int n,vector<vector<int>>&updates) {
        vector<int> diff(n,0);
        int low,high,c;
        for (int i=0;i<updates.size();i++) {
            low=updates[i][0];
            high=updates[i][1];
            c=updates[i][2];
            Update(diff,low,high,c);
        }
        return geta(diff);
    }
    void Update(vector<int>&diff,int low,int high,int c) {  //通过diff将A[low..high]增加c
        diff[low]+=c;
        if(high+1<diff.size())
            diff[high+1]-=c;
    }
    vector<int> geta(vector<int>&diff) {        //由差分数组diff构造ans
        int n=diff.size();
        vector<int> ans(n);
        ans[0]=diff[0];
        for(int i=1;i<n;i++)
            ans[i]=ans[i-1]+diff[i];
        return ans;
    }
};

File: vol1/ch11/LeetCode370.py
class Solution:
    def getModifiedArray(self,n:int,updates:List[List[int]]) -> List[int]:
        diff=[0]*n
        for i in range(0,len(updates)):
            low=updates[i][0];
            high=updates[i][1];
            c=updates[i][2];
            diff[low]+=c
            if high+1<n:diff[high+1]-=c
        ans=[0]*n
        ans[0]=diff[0]
        for i in range(1,n):
            ans[i]=ans[i-1]+diff[i]
        return ans
